### PR TITLE
http: Add non-owning `make_request` to http client

### PR DESCRIPTION
**Pull request by kreuzerkrieg**
To satisfy scylla's s3 client with future retry implementation add non-owning `make_request` to http client which is similar to the existing `make_request` but it doesn't take ownership of `request` and `reply_handler`

fixes: #2467

### CODE REVIEW

File: src/http/client.cc [hunk 1]
```diff
@@ -338,39 +338,45 @@ auto client::with_new_connection(Fn&& fn, abort_source* as) {
...
 
-            auto code = ex.code().value();
-            if ((code != EPIPE) && (code != ECONNABORTED)) {
-                return make_exception_future<>(ex);
-            }
+        if (!_retry) {
+            return make_exception_future<>(ex);
+        }
 
-            // The 'con' connection may not yet be freed, so the total connection
```
Comment (xemul) [inline src/http/client.cc:367]: There's a lot of indentation fixes here. When this happens, we ask to split the patch into two -- first, that does what it needs, but doesn't re-indent unaffected lines. The second patch restores the indentation. Please, do

File: include/seastar/http/client.hh [hunk 2]
```diff
@@ -295,7 +295,16 @@ public:
...
+    future<> make_request(request&& req, reply_handler&& handle, abort_source& as, std::optional<reply::status_type>&& expected = std::nullopt);
+
+    /**
+     * \brief Send the request and handle the response, same as \ref make_request()
+     *
+     *  @attention Note that the method does not take the ownership of the
+     * `request and the `handle`, it caller's responsibility the make sure they
+     * are referencing valid instances
+     */
+    future<> make_request(request& req, reply_handler& handle, abort_source& as, std::optional<reply::status_type> expected = std::nullopt);
```
Comment (xemul) [inline include/seastar/http/client.hh:307]: How about non-owning make_request that doesn't take abort-source argument?

File: src/http/client.cc (hunk 1 above)
Comment (kreuzerkrieg) [reply src/http/client.cc]: Sure. Done

File: include/seastar/http/client.hh (hunk 2 above)
Comment (kreuzerkrieg) [reply include/seastar/http/client.hh:309]: Added

File: include/seastar/http/client.hh [hunk 3]
```diff
@@ -295,7 +295,18 @@ public:
...
+    /**
+     * \brief Send the request and handle the response, same as \ref make_request()
+     *
+     *  @attention Note that the method does not take the ownership of the
+     * `request and the `handle`, it caller's responsibility the make sure they
+     * are referencing valid instances
+     */
+    future<> make_request(request& req, reply_handler& handle, std::optional<reply::status_type> expected = std::nullopt);
+
+    future<> make_request(request& req, reply_handler& handle, abort_source& as, std::optional<reply::status_type> expected = std::nullopt);
```
Comment (xemul) [inline include/seastar/http/client.hh:309]: This one should be documented as well

File: src/http/client.cc (hunk 1 above)
Comment (kreuzerkrieg) [reply src/http/client.cc]: agggr, squashed, lemme fix that

Comment (kreuzerkrieg) [reply src/http/client.cc:367]: done

File: include/seastar/http/client.hh (hunk 3 above)
Comment (kreuzerkrieg) [reply include/seastar/http/client.hh:316]: Done

**Pull request merged**